### PR TITLE
Added a small description how to debug with Eclipse

### PR DESCRIPTION
**Pull request by benjaminaigner**
Small addition to the docs, adding a step by step description how to use Eclipse for debugging.

### CODE REVIEW

**Comment by benjaminaigner** [comment]
By the way, your contribution guidelines are very funny, especially 
"Is the code adequately commented for people to understand how it is structured?".
Your own code is not really documented at all.

File: docs/openocd.rst [hunk 1]
```diff
@@ -129,3 +129,41 @@ resets the CPU cores, not the peripherals, which may lead to undefined behaviour
...
+    Select the option *Use remote target* and choose *Generic TCP/IP* as JTAG device. The hostname should be *localhost* and the port number
+    either *3333* (Core 0) or *3334* (Core 1).
+  * **Startup** Uncheck everything, except *Load image* and *Load symbols*. Both settings should be left to the default value (use project binary).
+  * **Common** Uncheck *Launch in background* and check *Debug* in *Display in favorites menu*
+
+After performing these steps, debugging with Eclipse should be possible. Please ensure
+to follow the next steps exactly in this order, otherwise it is possible to experience troubles. If something does not work,
+restart from the beginning.
+
+**Starting a debug session:**
```
Comment (projectgus) [inline docs/openocd.rst]: This is minor, but it'd be good thave this as a ReST header ie
Starting a debug sessions
~~~~~~~~~~~~~~~~~

**Comment by projectgus** [comment]
Hi Ben,

Thanks for sending this, much appreciated. I left a minor comment but it looks good.

> By the way, your contribution guidelines are very funny, especially
> "Is the code adequately commented for people to understand how it is structured?".
> Your own code is not really documented at all.

This is fair enough, commenting & internal developer documentation is a work in progress in several parts of the IDF code. If there are particular things which don't make sense and you don't feel they are adequately explained, please open Issues for them and we'll look at improving them.

File: docs/openocd.rst (hunk 1 above)
Comment (benjaminaigner) [inline docs/openocd.rst]: Should be done with d6bedea9

Regarding the other docs:
Yes, of course it is really hard work. It's one thing to have the software, but the docs are in most or many cases hard to keep up...
Anyway, thank you very much for this SDK!

**Comment by benjaminaigner** [comment]
> This is fair enough, commenting & internal developer documentation is a work in progress in several parts of the IDF code. If there are particular things which don't make sense and you don't feel they are adequately explained, please open Issues for them and we'll look at improving them

Yeah, I understand. It's quite hard to keep the docs up-to-date, especially if you also include 3rd party parts (Bluedroid, ...).
Anyway, thanks for the very useful IDF with all it's parts and drivers!

**Comment by ammaree** [comment]
Some feedback on this. I have followed the instructions exactly from the start but my configuration does not offer the  following options:
in **Debugger**
        no option for command set *Standard*
        no option for protocol version *mi*

When **Debug** session is started in Eclipse get error:

Error in final launch sequence
Failed to execute MI command:
load C:\\Dropbox\\devs\\ws\\esp32\\esp-1\\build\\application.elf 

Error message from debugger back end:
Load failed
Load failed

Any suggestions?

**Comment by benjaminaigner** [comment]
It might be related to the selected GDB launcher.

Did you select "Standard GDB Hardware Debugging Launcher" on the bottom of your Debug configuration?
If you use GDB (DSF) Hardware Debugging Launcher, no selection of the command set is possible.

**Comment by ammaree** [comment]
My config does not show the 2 options listed, only"GDB Hardware Debugging"
See attached screen snips

Andre

capture0

capture1

capture2

capture3

**Comment by benjaminaigner** [comment]
Ok, sry, I have not really an idea why you don't have this option.

My options:
screenshot from 2017-03-13 15-25-38

I have the possibility to select the launcher, because I installed GDB (DSF) Launcher too. Should not be the point, I selected the standard one.

Maybe try to install "C/C++ DSF GDB Debugger Integration" plugin, maybe it changes if you select the 
"Standard GDB Hardware Debugging Launcher", as it is shown on the bottom of the screenshot.

**Comment by benjaminaigner** [comment]
Another option to try:

Please put the provided launch file (it's compressed, please unzip it until esp32.launch is available) in following folder:

<workspace path>/.metadata/.plugins/org.eclipse.debug.core/.launches

Afterwards (maybe you need to restart eclipse), you should have the launch config. Or at least an error :-)

Adjust it to your binary settings & try to launch it.

Best wishes & good luck

**Comment by benjaminaigner** [comment]
Sry, no zip possible here?!?

**Comment by benjaminaigner** [comment]
Ok, no  zip, no xml, no password protected zip, but .txt :-)

Please download that file, rename it to esp32_test.launch
and proceed.


[esp32_test.txt](https://github.com/espressif/esp-idf/files/838600/esp32_test.txt)

**Comment by ammaree** [comment]
Downloaded, renamed, opened, changed the project name & executable name.
Options to configure remained the same
Error(s) remained the same..

**Comment by benjaminaigner** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

Is it possible for you to provide an extended debug output?

Either start it via console, or get the log file.

With further information, I might help you better.

On 13/03/17 16:23, André Maree wrote:
> Downloaded, renamed, opened, changed the project name & executable
> name. Options to configure remained the same Error(s) remained the
> same..
> 
> 

-----BEGIN PGP SIGNATURE-----

iQEzBAEBCAAdFiEEkZ15Sx/u8b4k40JmSXFOvY0eyL8FAljGuesACgkQSXFOvY0e
yL/49QgAl2rufJGsIDEmY3MDNtg+OGrZPneaXU7bs3XfqpSRLyPqYbbmqcu3OAW4
AsZxoMYehzGq/yV1Hs3Vna2KY2ie9DQ3skShTa/i1BCmUZ8orsbJsdBuX2omUiEU
Q4o4JRciDW6xp+EE9I56ngjW4VTVdrOIZo3qAeCNhmWllmnc54OcWm8Q7sIctU+T
hKyO2HUhQyZG3Roj8ImiSTlEQzup531gJ5KNctcajgE1dK+ftZdN23Oxn9RaFHz8
wDrfsSqiHqD4T+jHaCfh1b75kZO8tKCkpdiI+HGv911iDu6FUL7xFGrAQdmNW2IB
aOMwdfYvfNW86q3NaPY/+mW5N4vnrQ==
=icqc
-----END PGP SIGNATURE-----

**Comment by benjaminaigner** [comment]
@ammaree Maybe one additional hint: try to use the full absolute path for GDB command (in "Debugger").
I experienced problems, if I don't start Eclipse from the command line (Eclipse does not parse the .bashrc file).

**Comment by projectgus** [comment]
@ammaree did you end up making any more progress using gdb in Eclipse?

**Comment by ammaree** [comment]
@projectgus

Nope, no progress made in getting it to work. Have reverted back to printf() to do debugging until the OpenOCD/GDB/Eclipse integration improved.

ANY progress would be appreciated.

**Comment by benjaminaigner** [comment]
Hi,

did you try to start an Eclipse session via the terminal to provide additional log information?
Maybe I can help you, if you provide additional information from either Eclipse debug information or the terminal/commandline output.

**Comment by projectgus** [comment]
Hi @benjaminaigner ,

Sorry for the long delay between feedback.

As you've probably noticed, there's been a lot of work on openocd for esp32 lately plus some improvements in docs. The new multi-page guide to JTAG+openocd is here:
https://esp-idf.readthedocs.io/en/latest/api-guides/jtag-debugging/index.html

If you think there's anything missing or wrong in this guide, please feel free to open issues or send additional PRs.

Angus

**Pull request closed without merging**